### PR TITLE
Fixes Electrocutions to use available power, not consumed power

### PR DESCRIPTION
**Pull request by Varen**
## About the PR
Electrocution damage now scales with available power supply.

## Why / Balance
Electrocution damage used to scale based on power being used: An HV wire with 10GW available but only 100w being used would electrocute you for 100w worth of damage. It now uses the 10GW instead.
This means that cutting any wire with more than 70kw of available power will electrocute you. 
For reference, a single unmodified substation or SMES provides 150kw, APCs 10kw.

## Technical details
Scale damage off LastCombinedMaxSupply instead of LastCombinedSupply.

## Media
⚡

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
N/A

**Changelog**
:cl:
- fix: Cutting wires will now properly electrocute if enough power is available.

### CODE REVIEW

**Comment by TemporalOroboros** [comment]
Might have an issue with rad collectors instacritting people since IIRC they have a max supply of something like 1 billion.

**Comment by deltanedas** [comment]
and tesla rods which also use battery with big max output

**Comment by IProduceWidgets** [comment]
Oh my God thank you.

**Comment by TemporalOroboros** [comment]
Can you do me a favor and test this with the engine HV of a default 12 collector singulo setup? Post how much damage your character took, if it still exists afterwards.

**Comment by Anzarot121** [comment]
> Might have an issue with rad collectors instacritting people since IIRC they have a max supply of something like 1 billion.

I participated in a discussion of the PR, which reworked the electric shock so that it depended on the energy consumed and was not static, there is a maximum damage threshold.

**Comment by Varen** [comment]
> Might have an issue with rad collectors instacritting people since IIRC they have a max supply of something like 1 billion.

I have not changed how the damage calculation works. All this does is to make it consistent to available power rather than fluctuating with consumed.
In other words, the max damage you get now is the same as the max you'd get when the smes was fully uncharged before and you cut the wire to it. 

That said.. If you try to cut the hv to the singulo maybe you.. Kind of.. Should crit?

**Comment by PJB3005** [comment]
> Electrocution damage used to scale based on power being used: An HV wire with 10GW available but only 100w being used would electrocute you for 100w worth of damage. It now uses the 10GW instead.

That's not how it works. During the electrocution, an additional load representing the victim gets added to the power net. Looking at the code these are current 50 kW. This means you'll always have a minimum of 50 kW load and supply on the network.

**Comment by Varen** [comment]
> > Cutting wires will now properly electrocute if enough power is available.
> 
> Not true and also oversimplified. The problem of electrocution is complex, hence why a wall-of-text comment below.
> 
> I don't approve of this change. Electricity in SS14 approximates that of real life, and while there are several unknowns that I'm happy to compromise on, this change still leaves one huge inconsistency.
> 
> The "unknowns":
> 
>     * is the actor grounded (and how well)
> 
>     * voltage and current of the circuits involved
> 
>     * charge capacity of ground
> 
>     * charge capacity of actor's body
> 
>     * conductivity of actor's body
> 
>     * conductivity of wires
> 
> 
> For "gameplay's sake" I think it's fair to compromise on all of the above and just make axiomatic constant assumptions. So the way I'm seeing the current implementation of electrocution (prior to this PR) works like this:
> 
>     * the actor is never grounded
> 
>     * voltages and currents are omitted
> 
>     * when an electrocution happens, the actor becomes a "living wire" for the power net[1](#user-content-fn-1-95e18cc58bd4b7ee0ca354a3b8d9825d)
> 
> 
> Under these circumstances, the electrocution is caused by the current passing through the body of the victim. We don't keep track of current, but **consumed power** is a good approximation of it for gameplay purposes. The 50 kW load coming from "the victims body" can be interpreted as power turned into heat as the result of resistance of the victim's body and/or transference caused by partial grounding. Assumed constant for gameplay/simplicity purposes.
> 
> The new implementation of the electrocution (proposed in this PR) most closely approximates this:
> 
>     * the actor is always grounded
> 
>     * voltages and currents are omitted
> 
>     * ground capacity is infinite
> 
>     * conductivity of the actor's body is infinite
> 
>     * conductivity of the wires is infinite
> 
>     * when an electrocution happens, the actor becomes the short fuse for the power net[2](#user-content-fn-2-95e18cc58bd4b7ee0ca354a3b8d9825d)
> 
> 
> For the **available power** to play a role in electrocution, the only possible scenario is for the victim to close a circuit that is able to grab all of the supply's charge, a.k.a. **cause a short**. But there is a problem. If the charge passes through the victim's body, **where does it go**? In order to cause damage, it has to pass through victim's body (on its way to ground). This should result in all of the charge to be drained (SMES going down), and also the ground (which is presumably station's outer hull?) should receive all of this charge, and start blasting lightning at all surrounding grids. I don't see either of these mechanics happening in SS14 for various reasons (although it would be cool as fuck, admittedly).
> 
> Both of these "sets of working assumptions for an electrocution in SS14" technically work. But the "new model" slightly undermines physics by passing a current through victim's body that **comes from nowhere** and **goes to nowhere**, which I have to politely oppose.
> 
> I have to ask: what is the _**REAL**_ purpose of this PR? The formulas for electrocution can be adjusted basically freely since all the constants and multipliers used for calculations are changeable. You can adjust the damage/drain as you see fit **without** introducing more assumptions that go counter to real-life physics. What is the motivation behind making SS14 electrocutions approximate real life electrocutions _less accurately_?
> ### tl;dr:
> 
> There are several kinds of electrocutions IRL, and current SS14 electrocution model approximates one of them (human conductor[1](#user-content-fn-1-95e18cc58bd4b7ee0ca354a3b8d9825d)) quite nicely. This PR aims to change it so SS14 electrocution model approximates a different kind of electrocution (human short fuse[2](#user-content-fn-2-95e18cc58bd4b7ee0ca354a3b8d9825d)), but it does so poorly as there are numerous side effects to a short that current level of simulation in SS14 does not take into account. Would be really cool to see electric shorts being implemented in SS14 in the future, but **not like this**. The way it's done here is a bastardization of the concept, and I don't see _anything_ wrong with all SS14 electrocutions being the "human conductor"[1](#user-content-fn-1-95e18cc58bd4b7ee0ca354a3b8d9825d) kind.
> ## Footnotes
> 
>     1. For visualization: this is the kind of electrocution that happens IRL when a wire under load is split in two, falls into a puddle of water, and a person steps into that puddle. **Don't do this at home.** [↩](#user-content-fnref-1-95e18cc58bd4b7ee0ca354a3b8d9825d) [↩2](#user-content-fnref-1-2-95e18cc58bd4b7ee0ca354a3b8d9825d) [↩3](#user-content-fnref-1-3-95e18cc58bd4b7ee0ca354a3b8d9825d)
> 
>     2. For visualization: this is the kind of electrocution that happens IRL when you stick a fork in the power outlet, or grab the positive terminal of the circuit with one hand and the negative terminal with the other. **Don't do this at home.** [↩](#user-content-fnref-2-95e18cc58bd4b7ee0ca354a3b8d9825d) [↩2](#user-content-fnref-2-2-95e18cc58bd4b7ee0ca354a3b8d9825d)

Right now, with the current calculation, a player can use a wirecutter to cut the connection between the singulo and the SMES, because the SMES is full and isn't drawing enough power.
It's a simple matter of how intuitive the gameplay is: If I cut the HV between the singulo and the SMES, without insulated gloves, I should be fried. Right now that's not the case. Depending on the status of the smes and wires, you'll be able to just cut through it all with zero consequences. This is not the behaviour we want from wires, as it makes sabotage trivial.

If you'd like to suggest a different way to obtain that result, go ahead and suggest the relevant code changes. All I'm concerned about is that if you cut a wire that has gigawatts of energy available, with no insulation, you take damage commensurable to how bad of an idea that was.

**Comment by IamVelcroboy** [comment]
This is a wee bit over my head but does this fix/address [this](https://github.com/space-wizards/space-station-14/issues/24399) issue?

**Comment by Varen** [comment]
> It still needs the numbers checked - damage per common power source reviewed.
The damage should still be as per the table on #20880 , it's in the comments. Overall you shouldn't die from it unless the engineers heavily overbuilt the generators. Which.. fair.


> This is a wee bit over my head but does this fix/address [this](https://github.com/space-wizards/space-station-14/issues/24399) issue?

Yep, this is the same issue.

**Comment by IProduceWidgets** [comment]
This change is needed, or at least some change, that makes mapping shocked grilles not awful. Right now map's and mappers try to map shocked grills under windows and its incredibly non-trivial to figure out if they work. The information for players is even worse. 

For these reasons I support this PR, despite whatever real-life constraints it seems to violate. This is just a better game mechanic.

**Comment by deltanedas** [comment]
i think the 5kw is it will try to consume the load + 5kw, so it always charges
